frame/system/src/lib.rs

```diff
@@ -338,12 +321,15 @@ pub mod pallet {
 		/// functional/efficient alternatives.
 		type Lookup: StaticLookup<Target = Self::AccountId>;
 
-		/// The block header.
-		type Header: Parameter + traits::Header<Number = Self::BlockNumber, Hash = Self::Hash>;
+		/// The Block type used by the runtime. This is used by `construct_runtime` to retrieve the
+		/// extrinsics or other block specific data as needed.
+		#[pallet::no_default]
+		type Block: Parameter + Member + traits::Block<Hash = Self::Hash>;
 
 		/// Maximum number of block number to block hash mappings to keep (oldest pruned first).
 		#[pallet::constant]
-		type BlockHashCount: Get<Self::BlockNumber>;
+		#[pallet::no_default]
```

Comment:
Why is this the case?

Reply:
Seems like this is an issue with `derive_impl`. I see this error when I try to remove `pallet::no_default`:
`the trait bound `Self: pallet::Config` is not satisfied`. 
CC: @sam0x17. I will create an issue for this.

Reply:
https://github.com/paritytech/substrate/issues/14473

Reply:
@kianenigma how about this for a solution:
![image](https://github.com/paritytech/substrate/assets/1855021/70ede376-6dc9-4c45-a39a-2cd43e0310fb)

btw while we're at it, I would also love to change the syntax to be like:
```rust
#[register_default_trait(DefaultConfig)]
trait TestDefaultConfig {
    type Version = ();
    type BlockWeights = ();
    type BlockLength = ();
    type DbWeight = ();
    type Index = u64;
    #[no_disambiguation]
    type Block= SomeLocalTypeThatHopefullyExistsInThePallet;
    type Hash = sp_core::hash::H256;
    type Hashing = sp_runtime::traits::BlakeTwo256;
    type AccountId = AccountId;
    type Lookup = IdentityLookup<AccountId>;
    type BlockHashCount = frame_support::traits::ConstU64<10>;
    type AccountData = u32;
    type OnNewAccount = ();
    type OnKilledAccount = ();
    type SystemWeightInfo = ();
    type SS58Prefix = ();
    type MaxConsumers = frame_support::traits::ConstU32<16>;
}
```

This is what I meant when I said why not just use trait syntax and act like default associated types exist.... there is nothing stopping us from parsing something like the above since we're in an attribute macro

Reply:
This should allow us to provide defaults for a number of things we can't at the moment, the only caveat being they will be expected to be types that exist in the parent pallet, which I think is fine if there are only a few of them

Reply:
from discussion with kian: will probably call it `#[verbatim]`, `#[no_bounds]` or `#[unbounded]`

Reply:
I will have a look at the proposal here, but you should also take look at what is being done here before we proceed: https://github.com/paritytech/substrate/pull/14453